### DISCUSSION TITLE
Casting an array of PropMaps inside a PropMap

### DISCUSSION THREAD

**Discussion by grimsteel** [diwic/dbus-rs#464] [Q&A] [answered]
I have a PropMap (`a{sv}`). One of the keys in this PropMap, `foo`, maps to an array of PropMaps (`aa{sv}`). However, when I try to `cast` this to a `Vec<PropMap>`, it returns `None`.

Setup code:
```rust
// This is one of the PropMaps in the inner array
let mut inner_prop_map = PropMap::new();
inner_prop_map.insert("foobar".into(), Variant(Box::new(123)));

let prop_map_array = vec![inner_prop_map];

let mut prop_map = PropMap::new();
prop_map.insert("foo".into(), Variant(Box::new(prop_map_array)));
```

Doing the prop cast on `prop_map` directly works:
```rust
assert!(cast::<Vec<PropMap>>(prop_map.get("foo").unwrap()).is_some());
```

But when I send it into a `Message` and bring it back out, it doesn't work (in reality, I'm getting this PropMap as a return value from `org.freedesktop.NetworkManager` for `GetSettings`)

```rust
let msg = Message::new_method_call("a.b.c", "/a/b/c", "a.b.c", "AaBb").unwrap().append1(prop_map);
let another_map = msg.get1::<PropMap>().unwrap();
assert!(cast::<Vec<PropMap>>(another_map.get("foo").unwrap().is_none());
```

The only thing that works is using `as_iter` or `as_static_inner`. I'm not sure why casting doesn't work.

(I want to be able to use `cast_mut` to get a mutable reference to the array so I can edit specific fields on some of the inner dicts. Let me know if this isn't the right approach.)

**Comment by mulkieran** [comment]
@grimsteel I'm afraid I can not explain the behavior you've observed. I did notice that the last line of your last example:

```
assert!(cast::<Vec<PropMap>>(another_map.get("foo").unwrap().is_none());
```

is missing a closing paren.

**Reply by grimsteel** [reply]
Yes, it seems like my example code had a couple of typos.

This:
```rust
use dbus::{Message, arg::{PropMap, Variant, cast}};

// This is one of the PropMaps in the inner array
let mut inner_prop_map = PropMap::new();
inner_prop_map.insert("foobar".into(), Variant(Box::new(123)));

let prop_map_array = vec![inner_prop_map];

let mut prop_map = PropMap::new();
prop_map.insert("foo".into(), Variant(Box::new(prop_map_array)));

assert!(cast::<Vec<PropMap>>(&prop_map.get("foo").unwrap().0).is_some());

let msg = Message::new_method_call("a.b.c", "/a/b/c", "a.b.c", "AaBb").unwrap().append1(prop_map);
let inner_map = &another_map.get("foo").unwrap().0;

// The problem clearly shows itself here - the signature is
// an array of dicts of strings to variants, but it can't be casted to a vec of propmaps
println!("{:?}", inner_map.signature()); // Signature("aa{sv}\0")
assert!(cast::<Vec<PropMap>>(inner_map).is_none());
```

compiles and runs with all of the assertions passing.

**Reply by diwic** [reply, marked as answer]
Signatures that start with `aa` cannot be downcasted to, as they don't have a stable RefArg type.
 
This is kind of implied by the [argument guide](https://github.com/diwic/dbus-rs/blob/master/dbus/examples/argument_guide.md) 's reference section, but the docs could be clearer.

**Reply by grimsteel** [reply]
That makes more sense; I didn't really understand what "N/A" meant.

I can access the data inside with `as_iter` and `as_static_inner`, but is there any way to mutate the data inside of the array instead of copying the entire data structure?

**Reply by diwic** [reply]
@grimsteel I doubt it. I guess we have a few ways forward:
 1) Just clarify the docs. Copying data is likely pretty fast, comparing to other things libdbus and dbus-rs do anyway
 2) Add a `as_mut_iter` method, not sure how easy or hard this is. 
 3) Stabilize `aa{sv}` so there actually is a `Vec<PropMap>`. The implementation must be done carefully to avoid increased binary size (out of type explosion).

Any of these you feel like doing a PR for?

**Reply by diwic** [reply]
Added a sentence about "N/A" in the docs now.
